analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs

```diff
@@ -0,0 +1,144 @@
+        KnownType.System_Byte,
+        KnownType.System_SByte,
+        KnownType.System_Int16,
+        KnownType.System_UInt16,
+        KnownType.System_Int32,
+        KnownType.System_UInt32,
+        KnownType.System_Int64,
+        KnownType.System_UInt64,
+        KnownType.System_Char,
+        KnownType.System_Single,
+        KnownType.System_Double,
+        KnownType.System_Decimal,
+        KnownType.System_String,
+        KnownType.System_Object,
+        KnownType.Microsoft_AspNetCore_Http_IFormCollection,
+        KnownType.Microsoft_AspNetCore_Http_IFormFile,
+        KnownType.Microsoft_AspNetCore_Http_IFormFileCollection);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStart =>
+        {
+            if (!compilationStart.Compilation.ReferencesControllers())
+            {
+                return;
+            }
+            var examinedTypes = new ConcurrentDictionary<ITypeSymbol, bool>();
+
+            compilationStart.RegisterSymbolStartAction(symbolStart =>
+            {
+                var type = (INamedTypeSymbol)symbolStart.Symbol;
+                if (type.IsControllerType())
+                {
+                    symbolStart.RegisterSyntaxNodeAction(nodeContext => ProcessControllerMethods(nodeContext, examinedTypes), SyntaxKind.MethodDeclaration);
+                }
+            }, SymbolKind.NamedType);
+        });
+
+    private static void ProcessControllerMethods(SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        if (context.SemanticModel.GetDeclaredSymbol(context.Node) is IMethodSymbol method
+            && method.IsControllerMethod())
+        {
+            var modelParameterTypes = method.Parameters
+                .Select(x => x.Type)
+                .OfType<INamedTypeSymbol>()
+                .Where(x => !IgnoreType(x))
+                .SelectMany(x => RelatedTypesToExamine(x))
+                .Distinct();
+            foreach (var modelParameterType in modelParameterTypes)
+            {
+                CheckInvalidProperties(modelParameterType, context, examinedTypes);
+            }
+        }
+    }
+
+    private static void CheckInvalidProperties(INamedTypeSymbol parameterType, SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        var declaredProperties = new List<IPropertySymbol>();
+        GetAllDeclaredProperties(parameterType, examinedTypes, declaredProperties);
+        var invalidProperties = declaredProperties.Where(x => !CanBeNull(x.Type)
+            && !x.HasAttribute(KnownType.System_ComponentModel_DataAnnotations_RequiredAttribute)
+            && !x.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ModelBinding_Validation_ValidateNeverAttribute));
```

Comment:
Shouldn't this check be done in GetAllDeclaredProperties, so we avoid stepping into the type of such properties?
If so, please also add a test:
```cs
public class NotVisited
{
    public int Prop { get; set; } // Compliant
}
public class Model
{
    [ValidateNever]
    public NotVisited NotVisited { get; set; }
}
```

Reply:
[ValidateNever](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.modelbinding.validation.validateneverattribute?view=aspnetcore-8.0) can be applied to parameters, classes and properties.
So we also need support/tests for class/parameter:
```cs
public class MyController : Controller
{
    public void MyAction([ValidateNever] Model model) { }
}
```
and
```cs
[ValidateNever]
public class MyModel
{
    public int Prop { get; set; }
}
```